[SOAR 2311]New AD action modify object (#667)
rapid7/insightconnect-plugins: active_directory_ldap/help.md

  "host": "ldaps://example.com",
  "port": 389,
  "use_ssl": true,
  "username_password": "{\"username\":\"user1\", \"password\":\"mypassword\"}"
}
```

## Technical Details

### Actions

#### Modify Object

This action is used to modify an Active Directory objects attributes.

##### Input

|Name|Type|Default|Required|Description|Enum|Example|
|----|----|-------|--------|-----------|----|-------|
|attribute_to_modify|string|None|True|The name of the attribute to modify|None|postalCode|

> |attribute_value|string|None|True|The value of the attribute|None|1100|

Reviewer: If your example attribute is postalCode, shouldn't the example value be a valid postal code?

### Actions

#### Modify Object

> This action is used to modify an Active Directory objects attributes.

Reviewer: You can ignore this maybe, it reads weird to me. 

```suggestion
This action is used to modify the attributes of an Active Directory object.
```

Example output:

```
{
  "success": true
}
```

#### Add or Remove an Object from Group

This action is used to add or remove an object from an Active Directory group.

##### Input

|Name|Type|Default|Required|Description|Enum|Example|
|----|----|-------|--------|-----------|----|-------|
|add_remove|string|None|True|Add or remove the group|['add', 'remove']|add|
|distinguished_name|string|None|True|The distinguished name of the object whose membership will be modified|None|CN=user,OU=domain_users,DC=mydomain,DC=com|
|group_dn|string|None|True|The Distinguished Name of the group to add or remove|None|CN=group_name,OU=domain_groups,DC=example,DC=com|

Example input:

```
{
  "account_disabled": "true",

>   "additional_parameters": "{\"telephoneNumber\":\"(617)555-1234\"}",

Reviewer: Need to fix this

  "domain_name": "example.com",
  "first_name": "John",
  "last_name": "Doe",
  "logon_name": "jdoe",
  "password": "mypassword",
  "user_ou": "Users",
  "user_principal_name": "user@example.com"
}
```

##### Output

|Name|Type|Required|Description|
|----|----|--------|-----------|
|success|boolean|False|Operation status|

Example output:

> ```

Reviewer: Need example output.

```
{
  "account_disabled": "true",

>   "additional_parameters": "{\"telephoneNumber\":\"(617)555-1234\"}",

Reviewer: Need to fix the object in the example.

## Troubleshooting

Objects that contain an equals sign `=` or an asterisk `*` require the signs to be escaped.

> For example `CN=Robert = bob Smith,OU=domain_users,DC=rapid7,DC=com` must be escaped as`CN=Robert \= bob Smith,OU=domain_users,DC=mattsdomain,DC=local` in the input

Reviewer 2: ```suggestion
For example `CN=Robert = bob Smith,OU=domain_users,DC=rapid7,DC=com` must be escaped as`CN=Robert \= bob Smith,OU=domain_users,DC=mattsdomain,DC=local` in the input.
```

> A second example would be `CN=C**l guy,OU=domain_users,DC=rapid7,DC=com`must be escaped as `CN=C\*\*l guy,OU=domain_users,DC=rapid7,DC=com`

Reviewer 2: ```suggestion
A second example would be `CN=C**l guy,OU=domain_users,DC=rapid7,DC=com`. This must be escaped as `CN=C\*\*l guy,OU=domain_users,DC=rapid7,DC=com`.
```

> This plugin does not support objects an unpaired `\(\)` as part of their names.

Reviewer 2: ```suggestion
This plugin does not support objects and unpaired `\(\)` as part of their names.
```

> paired `\(\)` are supported

Reviewer 2: ```suggestion
Paired `\(\)` are supported.
```

For example `CN=Robert (Bob) Smith,OU=domain_users,DC=rapid7,DC=com` is supported

> but `CN=Robert Bob) Smith,OU=domain_users,DC=rapid7,DC=com` is not

Reviewer 2: ```suggestion
but `CN=Robert Bob) Smith,OU=domain_users,DC=rapid7,DC=com` is not.
```

objectname is the logon name of the user you are looking for. The DN can then be fed into another action by Repeating a collection for
the query results, and then using the variable step $item.dn

# Version History

> * 4.0.0 - New action Modify Object | Rename Modify Groups action to 'Add or Remove an Object from Group' | Fix issue where non ASCII characters were not being escaped

Reviewer 2: ```suggestion
* 4.0.0 - New action Modify Object | Rename Modify Groups action to 'Add or Remove an Object from Group' | Fix issue where non-ASCII characters were not being escaped
```
